test/e2e/run-openrpc-api-test-coverage.ts

```diff
@@ -0,0 +1,644 @@
+// it also validates that the JSON-RPC response is an error with
+// error code 4001 (user rejected request)
+class ConfirmationsRejectRule implements Rule {
+  private driver: any;
+  private only: string[];
+  private rejectButtonInsteadOfCancel: string[];
+  private requiresEthAccountsPermission: string[];
+
+  constructor(options: ConfirmationsRejectRuleOptions) {
+    this.driver = options.driver;
+    this.only = options.only;
+    this.rejectButtonInsteadOfCancel = [
+      'personal_sign',
+      'eth_signTypedData_v4',
+    ];
+    this.requiresEthAccountsPermission = [
+      'personal_sign',
+      'eth_signTypedData_v4',
+      'eth_getEncryptionPublicKey',
+    ];
+  }
+
+  getTitle() {
+    return 'Confirmations Rejection Rule';
+  }
+
+  async beforeRequest(_: any, call: Call) {
+    try {
+      if (this.requiresEthAccountsPermission.includes(call.methodName)) {
+        const requestPermissionsRequest = JSON.stringify({
+          jsonrpc: '2.0',
+          method: 'wallet_requestPermissions',
+          params: [{ eth_accounts: {} }],
+        });
+
+        await this.driver.executeScript(
+          `window.ethereum.request(${requestPermissionsRequest})`,
+        );
+        const screenshot = await this.driver.driver.takeScreenshot();
+        call.attachments = call.attachments || [];
+        call.attachments.push({
+          type: 'image',
+          data: `data:image/png;base64,${screenshot.toString('base64')}`,
+        });
+
+        await this.driver.waitUntilXWindowHandles(3);
+        await this.driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
+
+        await this.driver.findClickableElements({
+          text: 'Next',
+          tag: 'button',
+        });
+
+        const screenshotTwo = await this.driver.driver.takeScreenshot();
+        call.attachments.push({
+          type: 'image',
+          data: `data:image/png;base64,${screenshotTwo.toString('base64')}`,
+        });
+
+        await this.driver.clickElement({
+          text: 'Next',
+          tag: 'button',
+        });
+
+        await this.driver.clickElement({
+          text: 'Confirm',
+          tag: 'button',
+        });
+
+        await switchToOrOpenDapp(this.driver);
+      }
+    } catch (e) {
+      console.log(e);
+    }
+  }
+
+  async afterRequest(_: any, call: Call) {
+    try {
+      await this.driver.waitUntilXWindowHandles(3);
+      await this.driver.switchToWindowWithTitle(WINDOW_TITLES.Dialog);
+
+      let text = 'Cancel';
+      if (this.rejectButtonInsteadOfCancel.includes(call.methodName)) {
+        await this.driver.findClickableElements({
+          text: 'Reject',
+          tag: 'button',
+        });
+        text = 'Reject';
+      } else {
+        await this.driver.findClickableElements({
+          text: 'Cancel',
+          tag: 'button',
+        });
+      }
+      const screenshot = await this.driver.driver.takeScreenshot();
+      call.attachments = call.attachments || [];
```

Comment:
```suggestion
      call.attachments ??= [];
```

nit. feel free to ignore